.github/workflows/test.yml

```diff
@@ -0,0 +1,235 @@
+          - module: android
+            platform: Android
+          - module: webgl
+            platform: WebGL
+        exclude: ${{ fromJson(needs.base-and-hub.outputs.excludes) }}
+
+    steps:
+      ###########################
+      #          Setup          #
+      ###########################
+      - uses: actions/checkout@v2
+      - uses: docker/setup-qemu-action@v1
+      - uses: docker/setup-buildx-action@v1
+        with:
+          driver: docker
+      - name: Free disk space
+        run: .github/workflows/scripts/free_disk_space.sh
+
+      ###########################
+      #   Restore base and hub  #
+      ###########################
+      - uses: actions/download-artifact@v2
+      - name: Restore base and hub
+        run: |
+          docker load -i artifact/base-and-hub.tar
+          rm -rf artifact
+
+      ###########################
+      #      Setup to build     #
+      ###########################
+      - name: Setup to build
+        run: |
+          # The changeset for unity version
+          echo "CHANGESET=`npx unity-changeset ${{ matrix.version }}`" >> $GITHUB_ENV
+
+      ###########################
+      #       Build editor      #
+      ###########################
+      - name: Build
+        # if: steps.build-1.outcome == 'failure'
+        uses: docker/build-push-action@v2
+        id: build-1
+        continue-on-error: true
+        timeout-minutes: 40
+        with:
+          context: .
+          file: editor/Dockerfile
+          build-args: |
+            baseImage=base:dev
+            hubImage=hub:dev
+            version=${{ matrix.version }}
+            changeSet=${{ env.CHANGESET }}
+            module=${{ matrix.module }}
+          tags: editor:dev
+
+      ###########################
+      #   Build editor (retry)  #
+      ###########################
+      - name: Sleep for retry
+        if: steps.build-1.outcome == 'failure'
+        run: sleep 120
+
+      - name: Build (Retry)
+        if: steps.build-1.outcome == 'failure'
+        uses: docker/build-push-action@v2
+        id: build-2
+        # continue-on-error: true
+        timeout-minutes: 40
+        with:
+          context: .
+          file: editor/Dockerfile
+          build-args: |
+            baseImage=base:dev
+            hubImage=hub:dev
+            version=${{ matrix.version }}
+            changeSet=${{ env.CHANGESET }}
+            module=${{ matrix.module }}
+          tags: editor:dev
+
+      ###########################
+      #      Setup to test      #
+      ###########################
+      - name: Show image size
+        run: docker images
+      - name: Setup to test
+        run: |
+          # For 'linux-il2cpp' module, switch the script backend to 'IL2CPP'
+          if [ "${{ matrix.module }}" = "linux-il2cpp" ] ; then
+            mv -f reference-project-test/ProjectSettings/ProjectSettings_il2cpp.asset reference-project-test/ProjectSettings/ProjectSettings.asset
+          fi
```

Comment:
Note that altering the project will make the branch "dirty", meaning that builder will reject it unless "allowDirty" is passed. Maybe for testing purposes that would be ok.

Compared the changes between the il2cpp vs default project settings - note to self: 
![image](https://user-images.githubusercontent.com/20756439/104973950-a3ced480-59f6-11eb-8f66-9ece568384df.png)

@GabLeRoux this might be a nice way to reduce the amount of test projects; simply swap some settings only.

@mob-sakai do you think unity-builder should differentiate between these target platforms?

```yaml
          - module: base
            platform: StandaloneLinux64
          - module: linux-il2cpp
            platform: StandaloneLinux64
```

Reply:
@webbertakken 
In `unity-builder` and `unity-test-runner`, should we distinguish between the "StandaloneLinux64-Mono" platform and the "StandaloneLinux64-IL2CPP" platform?
Or do you mean to support the `base` and `linux-il2cpp` platforms in `unity-builder` and `unity-test-runner`?

Reply:
No, the way we [implemented](https://github.com/game-ci/unity-builder/pull/177/files#diff-e0ee836d4aea53d195cc8bd8bf5b6609a59da3e14f2df4518f3eac8724b1fc00R63-R69) it now is that StandaloneLinux64 (which doesn't have il2cpp until 2019.3) defaults to base and uses il2cpp from 2019.3 upwards.

Base is just a fallback because il2cpp doesn't exist, and we assumed that il2cpp would be the default/desired target platform for linux (and if not, that non-il2cpp would also be possible to build with il2cpp image).

I'm not sure what's best here. Suggestions?

Reply:
Ah, I understand what you mean.

Reply:
My Suggestion:
```
      - run: |
          # For 2019.3 or later, non-il2cpp would also be possible to build with il2cpp image.
          if [ `echo "${{ matrix.version }}" | grep -v '\(2018|2019.1|2019.2\)'` ] && [ "${{ matrix.module }}" = 'base' ] ; then
            echo "MODULE=linux-il2cpp" >> $GITHUB_ENV
          else
            echo "MODULE=${{ matrix.module }}" >> $GITHUB_ENV
          fi

      ###########################
      #       Build editor      #
      ###########################
      - name: Build
        # if: steps.build-1.outcome == 'failure'
        uses: docker/build-push-action@v2
        id: build-1
        continue-on-error: true
        timeout-minutes: 40
        with:
          context: .
          file: editor/Dockerfile
          build-args: |
            baseImage=base:dev
            hubImage=hub:dev
            version=${{ matrix.version }}
            changeSet=${{ env.CHANGESET }}
            module=${{ env.MODULE }}
          tags: editor:dev
```

Reply:
BTW, `unity-test-runner` seems to use `unityci/editor:<unity-version>-base-0` for testing.
https://github.com/game-ci/unity-test-runner/blob/main/src/model/image-tag.js

Reply:
Yea I like your suggestion, lets go for that!

> BTW, unity-test-runner seems to use unityci/editor:<unity-version>-base-0 for testing.

Yes, this is simply a result of our iterative approach. It should be able to test for all platforms to test platform-specific code paths. There's this [issue](https://github.com/game-ci/unity-test-runner/issues/89) that talks about custom image, but ideally we should also be able to use unity-test-runner in the same way as unity-builder, and specify targetPlatform.

Reply:
Fixed.